CRIU enable JDWP support post-restore (#1434)
eclipse-openj9/openj9-docs: docs/xrunjdwp.md

> ﻿<!--

Reviewer: > docs/xrunjdwp.md

`xrunjdwp` seems related to a specific legacy command option.
Suggest a generic name `criujdwp`. 
I expect this is a CRIU JDWP-specific doc instead of OpenJDK JDWP command options https://docs.oracle.com/en/java/javase/11/docs/specs/jpda/conninv.html.

Author: Yes, I had referred to this page to add the content. Please check and confirm

Author: As far as the file name `xrunjdwp.md` is concerned, it is according to the file naming convention that is followed in our documents. Since the name of the topic/option is `Xrunjdwp`, the name of the file should also be `xrunjdwp`.

* which accompanies this distribution and is available at
* https://www.eclipse.org/legal/epl-2.0/ or the Apache
* License, Version 2.0 which accompanies this distribution and
* is available at https://www.apache.org/licenses/LICENSE-2.0.
*
* This Source Code may also be made available under the
* following Secondary Licenses when the conditions for such
* availability set forth in the Eclipse Public License, v. 2.0
* are satisfied: GNU General Public License, version 2 with
* the GNU Classpath Exception [1] and GNU General Public
* License, version 2 with the OpenJDK Assembly Exception [2].
*
* [1] https://www.gnu.org/software/classpath/license.html
* [2] https://openjdk.org/legal/assembly-exception.html
*
* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
-->

# -Xrunjdwp

The `-Xrunjdwp` option enables the Java&trade; Debug Wire Protocol (JDWP) agent in the VM. The JDWP agent provides a connection point for the debugger to debug a Java application in that VM.

:fontawesome-solid-pencil:{: .note aria-hidden="true"} **Notes:**

> - You can enable JDWP support on Checkpoint/Restore In Userspace (CRIU) restore by specifying either the `-Xrunjdwp` or the `-agentlib:jdwp` command-line options pre-checkpoint.

Reviewer: There are three options to enable CRIU JDWP agent:
* -Xrunjdwp
* -agentlib:jdwp
* -agentpath:/path/to/libjdwp.so
The third option was missed in the initial https://github.com/eclipse-openj9/openj9-docs/issues/1407.

Author: Please provide the information for -agentpath:/path/to/libjdwp.so

Reviewer: > Please provide the information for -agentpath:/path/to/libjdwp.so

This is another way to enable CRIU JDWP via the JVMTI agent loading, more details of this command line option is at https://docs.oracle.com/javase/1.5.0/docs/guide/jvmti/jvmti.html#starting

> - Instead of specifying `-Xrunjdwp` or `-agentlib:jdwp` as command-line options pre-checkpoint when you restore a checkpoint image, you can add these command-line options by using the `CRIUSupport.registerRestoreOptionsFile` API.

Reviewer: ```suggestion
- Instead of specifying `-Xrunjdwp` or `-agentlib:jdwp` as command-line options pre-checkpoint when you restore a checkpoint image, you can add these command-line options into a restore option file and specify it via the `CRIUSupport.registerRestoreOptionsFile` API.
```

> - The `-Xrunjdwp` option is available both for OpenJDK 5 and later and for versions before OpenJDK 5.

Reviewer: Not sure if this is useful, Java 5 is long gone.
CRIU is supported on JDK 11+.

> - The `-agentlib:jdwp` option is available for OpenJDK 5 and later only. Although both the `-Xrunjdwp` and `-agentlib:jdwp` options are available for OpenJDK 5 and later, the `-agentlib:jdwp` option is preferable because the JDWP agent in OpenJDK 5 and later uses the Java Virtual Machine Tool Interface (JVMTI) interface to the VM rather than the older Java Virtual Machine Debug Interface (JVMDI) interface.

Reviewer: `-agentpath:/path/to/libjdwp.so` needs to be mentioned here as well.
Again, the Java 5 reference seems redundant.

## Parameters

### `address`

>         -Xrunjdwp:address=<value>

Reviewer: https://docs.oracle.com/en/java/javase/22/docs/specs/jpda/conninv.html#jdwpoptions has detailed suboption descriptions.

Author: Yes, I had. Please check and confirm. Thanks!

* is available at https://www.apache.org/licenses/LICENSE-2.0.
*
* This Source Code may also be made available under the
* following Secondary Licenses when the conditions for such
* availability set forth in the Eclipse Public License, v. 2.0
* are satisfied: GNU General Public License, version 2 with
* the GNU Classpath Exception [1] and GNU General Public
* License, version 2 with the OpenJDK Assembly Exception [2].
*
* [1] https://www.gnu.org/software/classpath/license.html
* [2] https://openjdk.org/legal/assembly-exception.html
*
* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
-->

# -Xrunjdwp

The `-Xrunjdwp` option enables the Java&trade; Debug Wire Protocol (JDWP) agent in the VM. The JDWP agent provides a connection point for the debugger to debug a Java application in that VM.

:fontawesome-solid-pencil:{: .note aria-hidden="true"} **Notes:**

- You can enable the JDWP support on Checkpoint/Restore In Userspace (CRIU) restore by specifying the following command-line options pre-checkpoint:

    - `-Xrunjdwp`
    - `-agentlib:jdwp`
    - `-agentpath:/path/to/libjdwp.so`

> - In OpenJ9, instead of specifying `-Xrunjdwp`, `-agentlib:jdwp`, or `-agentpath:/path/to/libjdwp.so` as command-line options pre-checkpoint, you can add these command-line options in a restore option file and specify it by using the `CRIUSupport.registerRestoreOptionsFile` API when you restore a checkpoint image.

Reviewer 2: Doesnt the enablement of jdwp via `CRIUSupport.registerRestoreOptionsFile` require `-XX:+debugonrestore` @JasonFengJ9

Reviewer: Yes, `jvmtiHookVMRestoreCRIUInit()` enables `jdwp` agent according to the option file settings, which is guarded by `isDebugOnRestoreEnabled(vm)` that requires `-XX:+debugonrestore`.
https://github.com/eclipse-openj9/openj9/blob/17ee5e7dd4e0b5b52321966907125d88aef13380/runtime/jvmti/jvmtiHook.c#L2050-L2051

`CRIUSupport.registerRestoreOptionsFile()` could register a restore option with `jdwp` along other command line options, but `jdwp` settings only effect with `-XX:+debugonrestore`.

Author: https://github.com/eclipse-openj9/openj9-docs/pull/1434#issuecomment-2438551654 - As per this comment, the following line was deleted:

> Restrictions: This option takes effect only when the -XX:+EnableCRIUSupport option and the -XX:+DebugOnRestore option are enabled.

Author: @tajila @JasonFengJ9 - Any changes to be made?

Reviewer 2: In that case I think we should only mention the capabiliteis that take effect without `-XX:+debugonrestore`. So I would omit anything that talks about adding jdwp options on restore.

Author: @tajila So, does that mean that the title that I have used in [What's new in version 0.48.0](version0.48.md#jdwp-support-on-checkpointrestore-in-userspace-criu-restore-is-enabled)  "JDWP support on Checkpoint/Restore In Userspace (CRIU) restore is enabled" is not valid? 

Looks like I might have to delete most of the content now. Could you please let me know what content I should delete and what I should retain? Thanks!

* is available at https://www.apache.org/licenses/LICENSE-2.0.
*
* This Source Code may also be made available under the
* following Secondary Licenses when the conditions for such
* availability set forth in the Eclipse Public License, v. 2.0
* are satisfied: GNU General Public License, version 2 with
* the GNU Classpath Exception [1] and GNU General Public
* License, version 2 with the OpenJDK Assembly Exception [2].
*
* [1] https://www.gnu.org/software/classpath/license.html
* [2] https://openjdk.org/legal/assembly-exception.html
*
* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
-->

# -Xrunjdwp

The `-Xrunjdwp` option enables the Java&trade; Debug Wire Protocol (JDWP) agent in the VM. The JDWP agent provides a connection point for the debugger to debug a Java application in that VM.

:fontawesome-solid-pencil:{: .note aria-hidden="true"} **Notes:**

- You can enable the JDWP support on Checkpoint/Restore In Userspace (CRIU) restore by specifying the following command-line options pre-checkpoint:

    - `-Xrunjdwp`
    - `-agentlib:jdwp`
    - `-agentpath:/path/to/libjdwp.so`

> - The main difference between `-Xrunjdwp` and `-agentlib:jdwp` is that `-Xrunjdwp` runs the JDWP agent as a separate process, while `-agentlib:jdwp` loads the JDWP agent as a Java library. That means that you can use `-Xrunjdwp` to debug a Java application that is running in a separate VM, and `-agentlib:jdwp` to debug a Java application that is running in the same VM. For more information, see [VM Invocation Options](https://docs.oracle.com/en/java/javase/22/docs/specs/jpda/conninv.html#vm-invocation-options).

Reviewer: The difference between `-Xrunjdwp` and `-agentlib:jdwp` was for historical reasons, I think OpenJ9 loads the same `jdwp` agent library eventually. 
Lines 36 & 37 can be removed.

﻿<!--
* Copyright (c) 2017, 2024 IBM Corp. and others
*
* This program and the accompanying materials are made
* available under the terms of the Eclipse Public License 2.0
* which accompanies this distribution and is available at
* https://www.eclipse.org/legal/epl-2.0/ or the Apache
* License, Version 2.0 which accompanies this distribution and
* is available at https://www.apache.org/licenses/LICENSE-2.0.
*
* This Source Code may also be made available under the
* following Secondary Licenses when the conditions for such
* availability set forth in the Eclipse Public License, v. 2.0
* are satisfied: GNU General Public License, version 2 with
* the GNU Classpath Exception [1] and GNU General Public
* License, version 2 with the OpenJDK Assembly Exception [2].
*
* [1] https://www.gnu.org/software/classpath/license.html
* [2] https://openjdk.org/legal/assembly-exception.html
*
* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
-->

# -Xrunjdwp

> The `-Xrunjdwp` option enables the Java&trade; Debug Wire Protocol (JDWP) agent in the VM. The JDWP agent provides a connection point for the debugger to debug a Java application in that VM.

Reviewer: `-agentlib:jdwp` enables the JDWP agent as per https://docs.oracle.com/en/java/javase/22/docs/specs/jpda/conninv.html#vm-invocation-options

### `suspend`

        -Xrunjdwp:suspend=[y|n]

: If `suspend=y`, the VM suspends running of the target Java application until the debugger is attached successfully and is ready for debugging. If `suspend=n`, the VM does not suspend running of the target Java application while the debug connection is established. The VM continues running the application as normal. The debugger can attach to the running application at any time.

: When a debug session is restored from a checkpoint, the VM suspends the running of the Java application that was being debugged even if `suspend=n`. The VM waits for the debugger to send a `Resume` request before it restarts the Java application for continuing the debug session. When the debug session is complete, the debugger disconnects from the JDWP agent, and the agent resumes normal running of the Java application.

### ![Start of content that applies to Java 11 (LTS) and later](cr/java11plus.png) `suspendOnRestore`

**(Linux&reg; x86, Linux on POWER&reg; (Little Endian), Linux on AArch64, and Linux on IBM Z&reg; only)**

This parameter is an OpenJ9 specific parameter.

        -Xrunjdwp:suspendOnRestore=[y|n]

> : If `suspendOnRestore=y`, the VM suspends the application on CRIU restore and debugging of the application does not happen. If `suspendOnRestore=n`, the VM does not suspend the application on restore, and the debug session is enabled.

Reviewer: `suspendOnRestore` is similar with `suspend`.
If `suspendOnRestore=y`, the VM suspends the running of the target Java application until the debugger is attached successfully and is ready for debugging upon restore.
If `suspendOnRestore=n`, the VM does not suspend the running of the target Java application while the debug connection is established upon restore.
This is a flag that only affects a VM post-restore.

> : If [`suspend=y`](#suspend) is set pre-checkpoint, the VM suspends the application pre-checkpoint, and on restore. The application remains suspended when you restore the VM even if you specify `suspend=n` pre-checkpoint. Therefore, to ensure that the application is not suspended on restore, you can use the `suspendOnRestore=n` setting to enable debugging after restore as well.

Reviewer: > you can use the `suspendOnRestore=n` setting to enable debugging after restore as well.

you can use the `suspendOnRestore=n` setting to NOT suspend the running of the JVM.

: The `suspendOnRestore` parameter is also useful if no debugging is set pre-checkpoint, but the debugging is enabled on restore.

> The debugger registers the post-restore event, receives notification upon CRIU restore, and suspends the target VM application unless `suspendOnRestore=n` is specified.

Reviewer: If no debugging was set pre-checkpoint, the debugging can be enabled via a restore option file and `CRIUSupport.registerRestoreOptionsFile` API.
The three command-line options that worked pre-checkpoint can be specified in the restore option file, and all sub-options including `suspendOnRestore` work similarly.
